pydoctor/templatewriter/__init__.py

```diff
@@ -56,130 +73,162 @@ def writeSummaryPages(self, system: System) -> None:
-                    return _HtmlTemplate(name=path.name, text=text)
-                else:
-                    return _StaticTemplate(name=path.name, text=text)
-        else:
-            warnings.warn(f"Cannot create Template: {path.as_posix()} is not recognized as template file. "
-                f"Template files must have one of the following extensions: {', '.join(cls.TEMPLATE_FILES_SUFFIX)}")
-        return None
-
-    def is_empty(self) -> bool:
-        """
-        Does this template contain nothing except whitespace?
-        Empty templates will not be rendered.
-        """
-        return len(self.text.strip()) == 0
+                template = Template.fromfile(entry, subdir=subdir)
+                if template:
+                    yield template
 
-    @abc.abstractproperty
-    def version(self) -> int:
+    @classmethod
+    def fromfile(cls, path: Union[Traversable, Path], subdir: Optional[PurePath] = None) -> Optional['Template']:
         """
-        Template version, C{-1} if no version.
-
-        HTML Templates should have a version identifier as follow::
-
-            <meta name="pydoctor-template-version" content="1" />
+        Create a concrete template object.
+        Type depends on the file extension.
 
-        This is always C{-1} for CSS and JS templates.
+        @param path: A L{Path} or L{Traversable} object that should point to a template file.
+        @param subdir: The subdirectory inside the template folder structure that holds this template file. 
+        @returns: The template object or C{None} if a the path entry is not a file.
+        @raises FailedToCreateTemplate: If there is an error while creating the template.
         """
-        raise NotImplementedError()
 
-    @abc.abstractproperty
-    def loader(self) -> Optional[ITemplateLoader]:
-        """
-        Object used to render the final file.
+        def suffix(name: str) -> str:
+            # Workaround to get a filename extension because
+            # importlib.abc.Traversable objects do not include .suffix property.
+            _, ext = splitext(name)
+            return ext
+        
+        def template_name(filename: str) -> str:
+            # The template name is the relative path to the template.
+            # Template files in subdirectories will have a name like: 'static/bar.svg'.
+            return subdir.joinpath(filename).as_posix() if subdir else filename
 
-        For HTML templates, this is a L{ITemplateLoader}.
+        if not path.is_file():
+            return None
 
-        For CSS and JS templates, this is C{None}
-        because there is no rendering to do, it's already the final file.
-        """
-        raise NotImplementedError()
+        file_extension = suffix(path.name).lower()
+        
+        template: Template
 
-class _StaticTemplate(Template):
+        try:
+            # Only try to decode the file text if the file is an HTML template
+            if file_extension == '.html':
```

Comment:
I'm not sure to see where we need the extra code to supporting case variations

Reply:
> we could just use .suffix since the relative path would have that property.

No we can't because Traversable object do not have .suffix property

Anyway, I've applied your comment regarding the compatness

Reply:
> I'm not sure to see where we need the extra code to support case variations

The constructor of `HtmlTemplate` is called with the file name in original case, so we'll later insert it into the dictionary in `TemplateLookup` in its original case as well, but we'll do lookups in that dictionary in lower case.

> > Or if we'd switch to base dir + relative path, we could just use .suffix since the relative path would have that property.
> 
> No we can't because Traversable object do not have .suffix property

With the base dir + relative path approach, the base dir would be `Traversable`, but the relative path would be a `PurePath` and we'd be checking the file type of the relative path.

Reply:
I see.

Yes maybe you approach is really needed after all.

Reply:
So I've changed it such that it takes a base dir + relative path to create the templates. 

I'm not sure what should be done about the casing of the template file extensions. The current behaviour doesn't look buggy to me

Reply:
We could import the [`CaseInsensitiveDict`](https://github.com/psf/requests/blob/master/requests/structures.py#L15) object from psf/requests project and use it to store the templates inside the TemplateLookup object.

This way we could not have two templates named `static/Img.svg` and `Static/Img.svg` (the later would override the first), but at the same time, the output files will be generated with the same casing as the template (`nameIndex.html` for instance)

Reply:
I've implemented **and tested** an alternative implementation using `CaseInsensitiveDict` here: https://github.com/tristanlatr/pydoctor/pull/1/files

Tell me if I should merge this PR into this branch. 

Thanks

Reply:
@mthuurne As we talked in private, I think case insensitive lookups is the expected behaviour.

I hope we can merge this code! Since it's pretty well tested :) 

Thanks

Reply:
Hi @mthuurne,

Please advise on the approach you would like!

Thanks

Reply:
Up!